ipa-core/src/protocol/hybrid/agg.rs

```diff
@@ -0,0 +1,232 @@
+    // if there are less, it's unattributed. if more, something went wrong.
+    for report in reports {
+        let match_key = report.match_key;
+        let entry = reports_by_matchkey.entry(match_key).or_insert((
+            [
+                AggregateableHybridReport::<BK, V>::ZERO,
+                AggregateableHybridReport::<BK, V>::ZERO,
+            ],
+            0,
+        ));
+        if entry.1 == 0 {
+            // If the count is 0, replace the first element with the new report
+            entry.0[0] = report.into();
+            entry.1 += 1;
+        } else if entry.1 == 1 {
+            // If the count is 1, replace the second element with the new report
+            entry.0[1] = report.into();
+            entry.1 += 1;
+        } else {
+            // If the count is 2 or more, increment the counter and drop the report
+            entry.1 += 1;
+        }
+    }
+
+    // we only keep the reports from match_keys that provided exactly 2 reports
+    let report_pairs: Vec<[AggregateableHybridReport<BK, V>; 2]> = reports_by_matchkey
+        .into_iter()
+        .filter_map(|(_, v)| if v.1 == 2 { Some(v.0) } else { None })
+        .collect::<Vec<_>>();
+
+    let chunk_size = TARGET_PROOF_SIZE;
+
+    let dzkp_validator = ctx.clone().dzkp_validator(
+        MaliciousProtocolSteps {
+            protocol: &HybridStep::GroupBySum,
+            validate: &HybridStep::GroupBySumValidate,
+        },
+        std::cmp::min(ctx.active_work().get(), chunk_size.next_power_of_two()),
+    );
+
+    let ctx = dzkp_validator
+        .context()
+        .set_total_records(TotalRecords::specified(report_pairs.len())?);
+
+    let agg_work = stream::iter(report_pairs)
+        .enumerate()
+        .map(|(idx, reports)| {
+            let agg_ctx = ctx.clone();
+            async move {
+                let (breakdown_key, _) = integer_add::<_, EightBitStep, 1>(
+                    agg_ctx.narrow(&AggregateReportsStep::AddBK),
+                    idx.into(),
+                    &reports[0].breakdown_key.to_bits(),
+                    &reports[1].breakdown_key.to_bits(),
+                )
+                .await?;
+                let (value, _) = integer_add::<_, EightBitStep, 1>(
+                    agg_ctx.narrow(&AggregateReportsStep::AddV),
+                    idx.into(),
+                    &reports[0].value.to_bits(),
+                    &reports[1].value.to_bits(),
+                )
+                .await?;
+                agg_ctx.validate_record(idx.into()).await?;
+                Ok::<_, Error>(AggregateableHybridReport::<BK, V> {
+                    match_key: (),
+                    breakdown_key: breakdown_key.collect_bits(),
+                    value: value.collect_bits(),
+                })
+            }
+        });
+
+    let agg_result = seq_join(ctx.active_work(), agg_work)
+        .try_collect::<Vec<_>>()
+        .await?;
+    Ok(agg_result)
+}
```

Comment:
stall fixed! thanks for the feedback @akoshelev 

now a new problem:
```
thread 'protocol::hybrid::agg::test::aggregate_reports_test' panicked at ipa-core/src/protocol/context/dzkp_validator.rs:937:32:
called `Result::unwrap()` on an `Err` value: ContextUnsafe("DZKPMaliciousContext")
```

it seems like this is related to validate not betting called, but it's called above on line 107. any suggestions?

Reply:
I can't spot any obvious problems with this code - validation is happening. It could be a miscalculation of batch size for malicious validation but I'd expect a stall in this case. @andyleiserson any thoughts why ZKP stuff may be throwing this?

Reply:
a clue: if I remove the validation on 107, it still panics in the same way...

the stacktrace isn't that helpful and doesn't seem to point where inside this function the panic is happening

Reply:
this happens when the validator is dropped, so it may be happening earlier than expected.

Reply:
do I need to validate for both steps?

Reply:
I added a second validate call and it didn't seem to help

Reply:
oh yea, I think you need to have a single validate after both steps

Reply:
tried that but it didn't help. (didn't push it because it didn't fix it)